typing/mtype.ml

```diff
@@ -175,17 +175,31 @@ let nondep_supertype env mid mty =
   in
     nondep_mty env Co mty
 
-let enrich_typedecl env p decl =
+let enrich_typedecl env p id decl =
   match decl.type_manifest with
     Some _ -> decl
   | None ->
       try
         let orig_decl = Env.find_type p env in
-        if orig_decl.type_arity <> decl.type_arity
-        then decl
-        else {decl with type_manifest =
-                Some(Btype.newgenty(Tconstr(p, decl.type_params, ref Mnil)))}
-      with Not_found ->
+        if decl.type_arity <> orig_decl.type_arity then
+          decl
+        else
+          let orig_ty =
+            Btype.newgenty(Tconstr(p, orig_decl.type_params, ref Mnil))
+          in
+          let new_ty =
+            Btype.newgenty(Tconstr(Pident id, decl.type_params, ref Mnil))
+          in
+          let env = Env.add_type ~check:false id decl env in
+          Ctype.mcomp env orig_ty new_ty;
```

Comment:
Is really `mcomp` the function you want to call here?
It is extremely conservative (only finds conflicts between types of different shapes, ignoring all nominal aspects).
The advantage is that it depends only very weakly on the environment, in particular it handles all abstract and missing types as wild-cards.
Note that there is no explicit guarantee that it will discover incompatibilities: it works only on a best effort basis. Soundness requires to be very careful when adding incompatibilities.
So the question is whether it is really sufficient for what you intend.

If you're going to use this function, you should probably add a comment about that, since originally it was designed to be used only during GADT pattern unification.

Reply:
OK, let's get back to the piece of code this is supposed to remplace.
Originally, the problem is that you could end up trying to find the type of a constructor with `get_constructor_type_path`, just to discover that it is a pair.
Since a pair is structurally incompatible with a sum type, this will be correctly detected.

In general, assuming that `id` is defined in `env` when trying to enrich the type declaration (otherwise it doesn't make sense), and that it points to a sum type definition, what `mcomp` tells you is that its original definition (scraping all privates) can only be a sum type (i.e. expands to a type path), an abstract type (also extends to a type path), or a type variable. So only this last case would be problematic, but I'm not sure how to build a counter-example for that: would adding `type 'a t = 'a` do the trick? Note also that if `env` is not the same calling `enrich_type_decl` and `get_constructor_type_path`, there might be other discrepancies.

Reply:
> would adding type 'a t = 'a do the trick

This raises a good point. My reasoning that we are making recursive modules as safe as GADTs ignores the fact that recursive modules can add higher-kinded equations -- which GADTs don't support. Possibly we need to extend `mcomp` to handle them as accurately as it handles ordinary types.